### PR TITLE
Dataset constuctor refactoring

### PR DESCRIPTION
**Pull request by Squire-tomsk**

### CODE REVIEW

File: datasets.py [hunk 1]
```diff
@@ -37,121 +40,124 @@ class AudioTextDataset(torch.utils.data.Dataset):
...
-		pop_meta = False,
-		string_array_encoding = 'utf_16_le',
-		_print = print,
-		debug_short_long_records_features_from_whole_normalized_signal=False
+			self,
+			data_paths: typing.List[str],
+			text_pipelines: typing.List[language_processing.ProcessingPipeline],
+			sample_rate: int,
+			frontend = None,
+			max_audio_file_size: float = None, ##MB
```
Comment (Squire-tomsk) [inline datasets.py]: add MB to var name

File: datasets.py [hunk 2]
```diff
@@ -37,121 +40,124 @@ class AudioTextDataset(torch.utils.data.Dataset):
...
 		self.audio_dtype = audio_dtype
 
 		data_paths = data_paths if isinstance(data_paths, list) else [data_paths]
-		exclude = set(exclude)
 
 		tic = time.time()
 
-		transcripts_read = list(map(transcripts.load, data_paths))
-		_print('Dataset reading time: ', time.time() - tic); tic = time.time()
+		segments = map(transcripts.load, data_paths)
```
Comment (Squire-tomsk) [inline datasets.py:85]: rename?

File: transcripts.py [hunk 3]
```diff
@@ -191,47 +195,93 @@ def sort(transcript):
...
+	align_boundary_words: bool = False,
+	cer: typing.Optional[Interval] = None,
+	wer: typing.Optional[Interval] = None,
+	mer: typing.Optional[Interval] = None,
+	duration: typing.Optional[Interval] = None,
+	gap: typing.Optional[Interval] = None,
+	num_speakers: typing.Optional[Interval] = None,
+	allowed_audio_names: typing.Set[str] = None,
+	allowed_unk_count: typing.Optional[Interval] = None,
+	groups: typing.Set = None,
```
Comment (vadimkantorov) [inline transcripts.py]: If `typing.Set` is untyped, pleas use `set` instead. Otherwise, please add a generic argument for the `set` element

File: datasets.py [hunk 4]
```diff
@@ -13,11 +13,15 @@
 import shaping
 import operator
 import typing
+import torch.nn.functional as F
+
+MB_SIZE = 1024 ** 2
```
Comment (vadimkantorov) [inline datasets.py]: unused?

File: datasets.py [hunk 5]
```diff
@@ -36,122 +40,137 @@ class AudioTextDataset(torch.utils.data.Dataset):
...
+			grouped_segments.extend(transcript)
+			transcripts_len.append(len(transcript))
+
+		_print('Dataset construction time: ', time.time() - tic)
+		tic = time.time()
+
+		self.bucket = torch.tensor(buckets, dtype = torch.short, device = 'cpu')
+		self.audio_path = utils.TensorBackedStringArray([t['audio_path'] for t in grouped_segments], encoding = string_array_encoding)
+		self.ref = utils.TensorBackedStringArray([t['ref'] for t in grouped_segments], encoding = string_array_encoding)
+		self.begin = torch.tensor([t['begin'] for t in grouped_segments], dtype = torch.float64, device = 'cpu')
```
Comment (vadimkantorov) [inline datasets.py]: are these still needed? we discussed that this information could remain in `meta` only

File: datasets.py [hunk 6]
```diff
@@ -273,16 +267,18 @@ def __getitem__(self, index):
...
 			transcript, speaker, features = transcript[0], speaker[0], features[0]
 			targets = [target[0] for target in targets]
 		return [transcript, speaker, features] + targets
 
 	def __len__(self):
 		return len(self.cumlen)
 
-	def collate_fn(self, batch) -> typing.Tuple[typing.List[dict], shaping.BS, shaping.BCT, shaping.B, shaping.BLY, shaping.B]:
-		if self.segmented:
+	# TODO проверить работу во всех режимах
```
Comment (vadimkantorov) [inline datasets.py]: just for review completeness, a TODO to fix before removing the DRAFT label

File: transcripts.py [hunk 7]
```diff
@@ -191,47 +195,93 @@ def sort(transcript):
...
+		ref = ' '.join(t['ref'].strip() for t in transcript)
+		speaker = []
+		for t in transcript:
+			speaker.append(
+				torch.full((len(t['ref']) + 1,), fill_value = t['speaker'], dtype = torch.int64, device = 'cpu')
+					 .scatter_(0, torch.tensor(len(t['ref'])), speaker_missing) # space handling
+			)
+		speaker: shaping._T = torch.cat(speaker)[:-1].unsqueeze(0) # [:-1] to drop last space, because of len(t['ref'] + 1)
+		assert len(ref) == speaker.shape[-1]
+		if all(t['speaker'] == transcript[0]['speaker'] for t in transcript):
```
Comment (vadimkantorov) [inline transcripts.py]: is this needed? normally, `collect_speaker_names` is supposed to fill out `speaker_name` to something reasonable

File: transcripts.py [hunk 8]
```diff
@@ -191,47 +195,93 @@ def sort(transcript):
...
+		audio_path = transcript[0]['audio_path']
+		assert all(t['audio_path'] == audio_path for t in transcript)
+		ref = ' '.join(t['ref'].strip() for t in transcript)
+		speaker = []
+		for t in transcript:
+			speaker.append(
+				torch.full((len(t['ref']) + 1,), fill_value = t['speaker'], dtype = torch.int64, device = 'cpu')
+					 .scatter_(0, torch.tensor(len(t['ref'])), speaker_missing) # space handling
+			)
+		speaker: shaping._T = torch.cat(speaker)[:-1].unsqueeze(0) # [:-1] to drop last space, because of len(t['ref'] + 1)
```
Comment (vadimkantorov) [inline transcripts.py]: As we discussed, you may move `unsqueeze` at a later point and use `_T` instead of `_T`

File: datasets.py [hunk 9]
```diff
@@ -330,11 +326,13 @@ def set_epoch(self, epoch):
 		rng.manual_seed(epoch)
 
 		def shuffle_and_split(g, batch_size):
-			required_samples_amount = math.ceil(len(g) / (batch_size * self.world_size)) * (batch_size * self.world_size)
-			extension_indices = torch.randint(0, len(g), size = (required_samples_amount - len(g),), generator = rng, device = g.device)
+			required_samples_amount = math.ceil(len(g) / (batch_size * self.world_size)) * (
```
Comment (vadimkantorov) [inline datasets.py]: nit: please rename "amount" into "num" or "count"
